### ISSUE TITLE
Problem when using IAsyncNotificationHandler

### ISSUE THREAD

**Issue by pjganesh** [jbogard/MediatR#202]
Version we are using 3.0.1.

Having the following classes:

```csharp
    class PayoutOrderRequestCommand : INotification { }
    class PayoutOrderRequestProcessHandler : IAsyncNotificationHandler<PayoutOrderRequestCommand>{ }
```
And the following code:

```csharp
container.RegisterSingleton<PayoutOrderRequestProcessHandler>();
container.RegisterCollection<IAsyncNotificationHandler<PayoutOrderRequestCommand>>(
                new[]
                {
                    typeof(PayoutOrderRequestProcessHandler)
                });

var command = new PayoutOrderRequestCommand(Guid.NewGuid(), correlationId);
var task = mediator.Publish(command, default(System.Threading.CancellationToken));
```

Getting below error when the Publish function is executed.

>No registration for type IEnumerable<INotificationHandler<PayoutOrderRequestCommand>> could be found.
------------------------------------------------------------------------------------

Please provide some solution for this issue.

**Comment by joejacobthomas** [comment]
we have the same problem using IAsyncNotificationHandler and i was planning to change the same to use IAsyncRequestHandler. Any help on this problem is greatly appreciated.

**Comment by pjganesh** [comment]
Can someone reply to this please ?

**Comment by stiankroknes** [comment]
Hi, 

From the question I assume SimpleInjector is used and the the default registration for _MultiInstanceFactory_ is used:
```
container.RegisterSingleton(new MultiInstanceFactory(container.GetAllInstances));
```
When Publish is executed the mediator internally calls the _MultiInstanceFactory_  (calling SimpleInjectors _GetAllInstances_) to resolve registered notification handlers - it tries to resolves in order _IAsyncNotificationHandler_, _INotificationHandler_ and last _ICancellableAsyncNotificationHandler_. 


A call to _GetAllInstances_ where no registration exists will cause _ActivationException_. Since SimpleInjector does not support asking if registration exists (that i know of) one could register the _MultiInstanceFactory_ as:
```
container.RegisterSingleton(new MultiInstanceFactory(type =>
{
    try  {  return container.GetAllInstances(type); }
    catch (ActivationException) { return Enumerable.Empty<object>(); }
}));
```

To summarize, the first resolve was fine, but when trying to resolve _INotificationHandler_ the _GetAllInstances_ throws _ActivationException_ since no registrations was found. Applying the above guard we should be able to resolve and publish successfully.

**Comment by pjganesh** [comment]
Thanks stiankroknes. Will try this out.

**Comment by abatishchev** [comment]
I believe you can check whether type is registered by using [`container.GetRegistration(type)`](https://simpleinjector.org/ReferenceLibrary/html/Overload_SimpleInjector_Container_GetRegistration.htm).

**Comment by rjperes** [comment]
The problem, as I see it, is: should MediatR throw an error if there are no handlers registered, or should it just ignore it? I believe the latter is more appropriate.

**Comment by jbogard** [comment]
The overall issue here is that MediatR is not explicitly in charge of "wiring up" handlers. In order to consolidate the API to just `Task<T> Send<T>(IRequest<T> request)`, I have to figure out which handlers actually exist by trying them one by one.

An alternative would be to consolidate handlers to *only* being async. Which I'm OK with honestly. It's annoying to have all these different handlers.

**Comment by abatishchev** [comment]
Personally would love to see single async and single sync handler type, and have no issues if the latter be adapted into the former, i.e. the library would call `Task.FromResult()` for you.

**Comment by jbogard** [comment]
The `IMediator` API is already async-only. But to determine if there's a sync or async handler, I have to attempt to resolve one or the other from the container. That can throw an exception.
